### PR TITLE
Convert the project to Swift 3.0 Xcode 8 b6

### PR DESCRIPTION
**Pull request by darthpelo**

### CODE REVIEW

**Comment by darthpelo** [comment]
Hi @tomkowz fell free to merge when Swift 3.0 will be official :)

**Comment by tomkowz** [comment]
Hi @darthpelo, thank you for your contribution. Great work!

File: Network/NetworkOperation.swift [hunk 1]
```diff
@@ -1,35 +1,35 @@
 import Foundation
 
-public class NetworkOperation: NSOperation {
+public class NetworkOperation: Operation {
     
     private var _ready: Bool
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:5]: Should be `_isReady`

File: Network/NetworkOperation.swift [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
-    public override var ready: Bool {
+    public override var isReady: Bool {
         get { return _ready }
         set { update({ self._ready = newValue }, key: "isReady") }
     }
     
     private var _executing: Bool
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:11]: Should be `_isExecuting`

File: Network/NetworkOperation.swift [hunk 3]
```diff
@@ -1,35 +1,35 @@
...
-    public override var executing: Bool {
+    public override var isExecuting: Bool {
         get { return _executing }
         set { update({ self._executing = newValue }, key: "isExecuting") }
     }
     
     private var _finished: Bool
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:17]: Should be `_isFinished`

File: Network/NetworkOperation.swift [hunk 4]
```diff
@@ -1,35 +1,35 @@
 import Foundation
 
-public class NetworkOperation: NSOperation {
+public class NetworkOperation: Operation {
     
     private var _ready: Bool
-    public override var ready: Bool {
+    public override var isReady: Bool {
         get { return _ready }
         set { update({ self._ready = newValue }, key: "isReady") }
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:8]: Should be `update(change:`

File: Network/NetworkOperation.swift [hunk 5]
```diff
@@ -1,35 +1,35 @@
...
     }
     
     private var _executing: Bool
-    public override var executing: Bool {
+    public override var isExecuting: Bool {
         get { return _executing }
         set { update({ self._executing = newValue }, key: "isExecuting") }
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:14]: Should be `update(change:`

File: Network/NetworkOperation.swift [hunk 6]
```diff
@@ -1,35 +1,35 @@
...
     }
     
     private var _finished: Bool
-    public override var finished: Bool {
+    public override var isFinished: Bool {
         get { return _finished }
         set { update({ self._finished = newValue }, key: "isFinished") }
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:20]: Should be `update(change:`

File: Network/NetworkOperation.swift [hunk 7]
```diff
@@ -1,35 +1,35 @@
...
     }
     
     private var _cancelled: Bool
-    public override var cancelled: Bool {
+    public override var isCancelled: Bool {
         get { return _cancelled }
         set { update({ self._cancelled = newValue }, key: "isCancelled") }
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:26]: Should be `update(change:`

File: Network/NetworkOperation.swift [hunk 8]
```diff
@@ -1,35 +1,35 @@
...
     }
     
     private var _finished: Bool
-    public override var finished: Bool {
+    public override var isFinished: Bool {
         get { return _finished }
         set { update({ self._finished = newValue }, key: "isFinished") }
     }
     
     private var _cancelled: Bool
```
Comment (Fawxy) [inline Network/NetworkOperation.swift:23]: Should be `_isCancelled`

File: Network/NetworkService.swift [hunk 9]
```diff
@@ -2,32 +2,41 @@ import Foundation
...
     
-    private var task: NSURLSessionDataTask?
-    private var successCodes: Range<Int> = 200..<299
-    private var failureCodes: Range<Int> = 400..<499
+    private var task: URLSessionDataTask?
+    private var successCodes: CountableRange<Int> = 200..<299
+    private var failureCodes: CountableRange<Int> = 400..<499
     
     enum Method: String {
         case GET, POST, PUT, DELETE
```
Comment (Fawxy) [inline Network/NetworkService.swift:10]: Enum should be lowercase

File: Network/NetworkService.swift [hunk 10]
```diff
@@ -2,32 +2,41 @@ import Foundation
...
     }
     
-    func request(url url: NSURL, method: Method,
+    enum QueryType {
+        case JSON, PATH
```
Comment (Fawxy) [inline Network/NetworkService.swift:14]: Enum should be lowercase

File: Network/NetworkService.swift [hunk 11]
```diff
@@ -63,4 +72,34 @@ class NetworkService {
     func cancel() {
         task?.cancel()
     }
+    
+    
+    //MARK: Private
+    private func makeQuery(for url: URL, params: [String: AnyObject], type: QueryType) -> NSMutableURLRequest? {
```
Comment (Fawxy) [inline Network/NetworkService.swift:78]: Should return a `URLRequest`

File: Network/NetworkService.swift [hunk 12]
```diff
@@ -63,4 +72,34 @@ class NetworkService {
...
+        switch type {
+        case .JSON:
+            let mutableRequest = NSMutableURLRequest(url: url, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
```
Comment (Fawxy) [inline Network/NetworkService.swift:81]: should be `var mutableRequest = URLRequest(url:`

File: Network/NetworkService.swift [hunk 13]
```diff
@@ -63,4 +72,34 @@ class NetworkService {
...
+                query = query + key + "=" + (value as! String) + "&"
+            }
+            
+            var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
+            components?.query = "\(query)"
+            
+            guard let comp = components else {
+                return nil
+            }
+            return NSMutableURLRequest(url: comp.url!, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData, timeoutInterval: 10.0)
```
Comment (Fawxy) [inline Network/NetworkService.swift:99]: Should be `URLRequest(url:`

**Review by Fawxy — CHANGES_REQUESTED** [review]
Few things to make this properly Swift 3-ified. Should also be noted that `JSONSerialization` now returns `Any` rather than `AnyObject`.

File: Network/NetworkService.swift (hunk 11 above)
Comment (tomkowz) [reply Network/NetworkService.swift:78]: Later on you need to set header fields and http method so IMO it has to be NSMutableURLRequest.

Also, I would say it should return non-optional. If something went wrong in here it means that request was build incorrectly and this is a programmer bug that has to be fixed. Any thoughts why it should return optional?

Comment (Fawxy) [reply Network/NetworkService.swift:78]: `URLRequest` is mutable provided it is not defined as a `let` constant

File: Network/NetworkService.swift (hunk 12 above)
Comment (tomkowz) [reply Network/NetworkService.swift:81]: You can extract cache policy and timeout interval before switch and use them in this switch.

File: Network/NetworkService.swift (hunk 11 above)
Comment (tomkowz) [reply Network/NetworkService.swift:78]: How is that possible? documentation clearly says `allHTTPHeaderFields` is read-only: https://developer.apple.com/reference/foundation/nsurlrequest/1418477-allhttpheaderfields

`httpMethod` too.

Could you share a working example if it works as you said?

Comment (Fawxy) [reply Network/NetworkService.swift:78]: Sure - give me a sec to fork :) there are a bunch of other changes needed too for the final release of Xcode 8

Comment (tomkowz) [reply Network/NetworkService.swift:78]: Cool, thank you!

**Comment by tomkowz** [comment]
Hey @darthpelo. Thank you for your work again. Great job! We took your piece of code to the https://github.com/tomkowz/NetworkLayerExample/pull/3 PR and improved it a bit.

**Comment by darthpelo** [comment]
@tomkowz glad to help you :)

**Pull request closed without merging**